### PR TITLE
make istiod-less as default remote for integration test

### PR DESCRIPTION
**Pull request by irisdingbj**

### CODE REVIEW

File: manifests/charts/istio-control/istio-discovery/templates/configmap.yaml
Comment (stevenctl) [inline manifests/charts/istio-control/istio-discovery/templates/configmap.yaml:54]: Why not just fall-though to the one below? Is there a reason we wouldn't want `istiod-revision.istio-system.svc`?

File: pkg/test/framework/components/istio/operator.go
Comment (stevenctl) [inline pkg/test/framework/components/istio/operator.go:403]: We should eventually add some methods to `resource.Cluster` like `IsConfigCluster` rather than leaving that on the env. If we had that, we wouldn't have to change the signature here, and we could just check that inside of `generateCommonInstallSettings`. Not needed in this PR, just something I've meant to add for a while.

File: manifests/charts/istio-control/istio-discovery/templates/configmap.yaml
Comment (irisdingbj) [reply manifests/charts/istio-control/istio-discovery/templates/configmap.yaml:54]: when we are in remote cluster,  we will create `istiod` or `istiod-remote` endpoint.  gateways will get `discoveryAddress` from the def here.

File: pkg/test/framework/components/istio/operator.go
Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go:404]: yep, Let me clear this PR out first. We can address this later.

**Comment by irisdingbj** [comment]
blocked by https://github.com/istio/istio/issues/28220 and https://github.com/istio/istio/issues/28218

File: tests/integration/iop-remote-integration-test-defaults.yaml
Comment (linsun) [inline tests/integration/iop-remote-integration-test-defaults.yaml:9]: do we want to test this set as true?  We are not enabling this by default in our MC docs.  Could you double check?

Comment (stevenctl) [reply tests/integration/iop-remote-integration-test-defaults.yaml:9]: This was set as true in #26995, I think it's good to test with it on so we can work on stabilizing it.

File: manifests/charts/istio-control/istio-discovery/templates/configmap.yaml
Comment (stevenctl) [reply manifests/charts/istio-control/istio-discovery/templates/configmap.yaml]: Right, but we could have `if and (.Values.global.remotePilotAddress .Values.pilot.enabled` resulting in istiod-remote, because we already have a branch that gives `discoveryAddress: istiod`.

If we're doing this in a separate branch because we don't want to include the revision intentionally, can we call that out with a comment?

Comment (irisdingbj) [reply manifests/charts/istio-control/istio-discovery/templates/configmap.yaml]: yes, This is to align the endpoint def in https://github.com/istio/istio/blob/master/manifests/charts/base/templates/endpoints.yaml#L2

you see we actually will create `istiod` or `istiod-remote`

**Comment by irisdingbj** [comment]
@stevenctl Thanks for the hint about remote secret.  after I created remote secret for all clusters testing passed now.

@stevenctl @linsun @nmittler PTAL, Thanks.

File: pkg/test/framework/components/istio/operator.go
Comment (stevenctl) [inline pkg/test/framework/components/istio/operator.go:744]: How can we make this more like what a user would do? I don't imagine we'll have them create endpoints/service by hand.

File: pkg/test/framework/components/istio/operator.go
Comment (stevenctl) [inline pkg/test/framework/components/istio/operator.go:669]: Can we make this conditional on what kind of cluster? We don't want to run into a scenario where a (non-config) remote cluster passes in CI, but fails in an actual deployment due to permissions issues.

File: pkg/test/framework/components/istio/operator.go
Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go:727]: added `"--set", "values.global.remotePilotAddress="+remoteIstiodAddress.IP.String()` to let it create the istiod endpoints automatically. Now this part is not used.

File: pkg/test/framework/components/istio/operator.go
Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go:652]: We will create these serviceAccounts anyway if they do not exist as the PR in https://github.com/istio/istio/pull/28059

Comment (stevenctl) [reply pkg/test/framework/components/istio/operator.go:652]: This constant is `istiod-service-account`. By default this command sets things up to use `istio-reader`. If features are developed such that they require a permission that `istiod-service-account` has but `istio-reader` does not, integration tests will pass and users will fail since we don't document that they need to switch which service account to use.

File: pkg/test/framework/components/istio/operator.go
Comment (stevenctl) [reply pkg/test/framework/components/istio/operator.go:727]: Are we able to delete this method then?

Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go:727]: no.   As external Istiod test case is still relying on it.  ( for external Istiod case, we will install config cluster first and in that case it is not aware of the control plane istiod address, so "--set", "values.global.remotePilotAddress="+remoteIstiodAddress.IP.String()" will not work.)

File: pkg/test/framework/components/istio/operator.go
Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go:652]: The issue is  istio-reader  lack one authority.  details see: https://github.com/istio/istio/pull/28389

after that PR is merged, we are safe to switch back to `istio-reader-service-accout` as the default one.

Comment (stevenctl) [reply pkg/test/framework/components/istio/operator.go:652]: My point is: we should use istio reader for remotes and istiod-... for config cluster.

File: tests/integration/iop-remote-integration-test-defaults.yaml
Comment (linsun) [reply tests/integration/iop-remote-integration-test-defaults.yaml:9]: i see thanks - okay to have this to true for master.  if this is cherry-picked to 1.8, I think it would be good to test what we want users to do.

File: tests/integration/iop-remote-integration-test-defaults.yaml
Comment (linsun) [inline tests/integration/iop-remote-integration-test-defaults.yaml:1]: should we use remote profile here in the test, similarly as what we tell users to do?

File: pkg/test/framework/components/istio/operator.go
Comment (linsun) [reply pkg/test/framework/components/istio/operator.go:727]: yep, so the challenge is we sometimes don't know istiod's endpoint ahead of time in our test framework.  If we do, we should not have this issue.  

I understand we have this issue for external istiod as it needs to set up config cluster first, but for MC primary-remote, you could setup primary first, then remote which will have the remoteIstiodAddress.   Is it possible not to use this path for MC primary-remote test?

File: tests/integration/iop-remote-integration-test-defaults.yaml
Comment (irisdingbj) [reply tests/integration/iop-remote-integration-test-defaults.yaml:1]: remote profile is empty for now. 
```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec: {}
```
Should we update it to use the version like here?

Comment (linsun) [reply tests/integration/iop-remote-integration-test-defaults.yaml:1]: Yes i was surprised to find this remote profile is wrong.  I think it is a just placeholder now. should be a profile that has pilot disabled.

Comment (linsun) [reply tests/integration/iop-remote-integration-test-defaults.yaml:1]: ok to address this comment as a follow up PR. :)

File: pkg/test/framework/components/istio/operator.go
Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go]: yes, after above PR is merged, we will use istio reader for remote.

**Comment by irisdingbj** [comment]
depends on #28389

**Comment by irisdingbj** [comment]
@nmittler  @stevenctl @linsun PTAL , Thanks!

File: pkg/test/framework/components/istio/config.go
Comment (nmittler) [inline pkg/test/framework/components/istio/config.go:113]: Could we make it `ConfigClusterIOPFile` and `RemoteClusterIOPFile` to be more clear?

File: pkg/test/framework/components/istio/operator.go
Comment (nmittler) [inline pkg/test/framework/components/istio/operator.go:63]: maybe just make it a minute?

File: pkg/test/framework/components/istio/operator.go
Comment (nmittler) [inline pkg/test/framework/components/istio/operator.go:248]: Why have we gotten rid of the concurrency here? IIUC this was done intentionally to decrease startup time.

FYI @stevenctl

File: pkg/test/framework/components/istio/operator.go
Comment (nmittler) [inline pkg/test/framework/components/istio/operator.go:279]: Could we just always do this, not just for multicluster? We want this to only apply the remote secret to `other` cluster anyway, so this should be a no-op for single cluster.

Generally, I'd like to get rid of the `if env.IsMulticluster()` blocks to the extent possible ... it shouldn't be a special case.

File: pkg/test/framework/components/istio/operator.go
Comment (nmittler) [inline pkg/test/framework/components/istio/operator.go:283]: This condition looks wrong to me. Should it be `if env.IsRemoteCluster(cluster)`?

File: pkg/test/framework/components/istio/operator.go
Comment (nmittler) [inline pkg/test/framework/components/istio/operator.go:286]: Should we just always deploy the east-west gateway in all cluster?

@stevenctl wdyt?

File: pkg/test/framework/components/istio/operator.go
Comment (nmittler) [inline pkg/test/framework/components/istio/operator.go:296]: Let's rename this `i.exposeUserServices(cluster)`. The name is not immediately clear.

File: pkg/test/framework/components/istio/operator.go
Comment (nmittler) [reply pkg/test/framework/components/istio/operator.go:725]: @irisdingbj `config cluster` != `remote cluster`. The installation of a config cluster should have no reference to istiod (i.e. does not use `values.global.remotePilotAddress`). We should safely be able to install clusters as follows:

1. Install config clusters (basic needed for external istiod)
1. Install primary clusters
1. Install remote clusters (may or may not be the same clusters as the config clusters)

I think we should change this here in this PR unless there is an overwhelming reason why it can't be done now.

File: pkg/test/framework/components/istio/operator.go
Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go:248]: the concurrency is added recently by @stevenctl ,  Let me consume it here to keep the concurrency.

File: pkg/test/framework/components/istio/operator.go
Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go:283]: We do not have `IsRemoteCluster` method defined..  Let me add a one in env

File: pkg/test/framework/components/istio/operator.go
Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go:296]: yes, agree this will align our multi-cluster doc as well.

File: pkg/test/framework/components/istio/operator.go
Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go:286]: according to our doc, we only deploy this for different network.

File: pkg/test/framework/components/istio/operator.go
Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go:279]: removed

File: pkg/test/framework/components/istio/config.go
Comment (irisdingbj) [reply pkg/test/framework/components/istio/config.go:113]: fixed

File: pkg/test/framework/components/istio/operator.go
Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go:727]: yes, now only remote cluster needs `values.global.remotePilotAddress`. Now for externalIstiod test case, we are using the same cluster for both config and remote cluster.  We can modify this later to use 3 clusters to distinguish between config and remote cluster.

Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go:727]: yes, now only remote cluster needs values.global.remotePilotAddress. Now for externalIstiod test case, we are using the same cluster for both config and remote cluster. We can modify this later to use 3 clusters to distinguish between config and remote cluster.

File: pkg/test/framework/components/istio/operator.go
Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go]: fixed

File: pkg/test/framework/components/istio/operator.go
Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go]: Error patching Webhook Get "https://127.0.0.1:34611/apis/admissionregistration.k8s.io/v1beta1/mutatingwebhookconfigurations/istio-sidecar-injector": dial tcp 127.0.0.1:34611: connect: connection refused .   we can not create remote secret if it is only single cluster.. I turned it back.

**Comment by irisdingbj** [comment]
@nmittler PTAL for the latest change. thanks.

File: pkg/test/framework/components/istio/operator.go
Comment (stevenctl) [inline pkg/test/framework/components/istio/operator.go:541]: In a real-world install would we have to specify these in `set` flags? Can you leave a comment that we either plan to clean this up, or are doing this as a workaround for our test env.

Comment (irisdingbj) [reply pkg/test/framework/components/istio/operator.go:541]: yes, this is needed even in real-world in current code base.  So when user install remote cluster, they need set this via IOP or --set after they get the eastwest GW address in primary cluster.  

So  this is not a workaround.  If we plan to clean this up, we may need a seperate design doc to discuss how to land this.

Comment (linsun) [reply pkg/test/framework/components/istio/operator.go:545]: one proposal is to generate the url based on discoverAddr and cluster and network info, which we have all that info from other places.  I agree we should look into simplify this post 1.8.  I'll open a WI to track it.

File: pkg/test/framework/components/istio/operator.go
Comment (linsun) [reply pkg/test/framework/components/istio/operator.go:728]: Agree config cluster doesn't need remotePilotAddr.  Only discoverAddr

**Review by linsun — COMMENTED** [review]
@irisdingbj the change LG to me.  I'll give a bit time for @nmittler or @stevenctl to review today.

File: pkg/test/framework/components/istio/operator.go
Comment (nmittler) [inline pkg/test/framework/components/istio/operator.go:64]: I wonder if you needed to increase this because you had previously removed the concurrency when installing? Can you verify that you still need this change?

File: pkg/test/framework/components/istio/operator.go
Comment (nmittler) [inline pkg/test/framework/components/istio/operator.go:488]: Isn't discovery only needed for remote clusters? Config clusters should not use discovery.

File: pkg/test/framework/components/istio/operator.go
Comment (nmittler) [reply pkg/test/framework/components/istio/operator.go:728]: It shouldn't need either `discoveryAddr` or `remotePilotAddr`. The config cluster is inherently the first thing that should be configured because the control plane relies on it (not the other way around). We have to fix this before 1.9.

File: tests/integration/iop-integration-test-defaults.yaml
Comment (nmittler) [inline tests/integration/iop-integration-test-defaults.yaml:78]: Looks like you removed the trailing carriage return.

**Review by nmittler — APPROVED** [review]
@irisdingbj I've approved ... I'm fine just getting this in and iterating. We have a lot of things to clean up in general.

**Comment by irisdingbj** [comment]
> @irisdingbj I've approved ... I'm fine just getting this in and iterating. We have a lot of things to clean up in general.

Thanks @nmittler Let me address your new comments in a following up PR.

**Pull request merged**
